src/Native/Unix/System.Native/pal_io.cpp

```diff
@@ -364,6 +365,60 @@ extern "C" int32_t SystemNative_ReadDirR(DIR* dir, void* buffer, int32_t bufferS
     return 0;
 }
 
+#else
+// To reduce the number of string copies, this function calling pattern works as follows:
+// 1) The managed code calls GetDirentSize() to get the platform-specific
+//    size of the dirent struct.
+// 2) The managed code creates a byte[] buffer of the size of the native dirent
+//    and passes a pointer to this buffer to this function.
+// 3) This function gets a pointer to the possibly statically allocated directory entry.
+// 4) Then, byte[] entry is copied into the byte[] buffer for bufferSize bytes.
+//    This makes the 1st strcpy.
+// 5) The ConvertDirent function will set a pointer to the start of the inode name
+//    in the byte[] buffer so the managed code can find it and copy it out of the
+//    buffer into a managed string that the caller of the framework can use, making
+//    the 2nd and final strcpy.
+//
+//    To make this function thread safe ensure calls on the same DIR* dir never happen concurrently.
+extern "C" int32_t SystemNative_ReadDirR(DIR* dir, void* buffer, int32_t bufferSize, DirectoryEntry* outputEntry)
+{
+    assert(buffer != nullptr);
+    assert(dir != nullptr);
+    assert(outputEntry != nullptr);
+
+    if (bufferSize < static_cast<int32_t>(sizeof(dirent)))
+    {
+        assert(false && "Buffer size too small; use GetDirentSize to get required buffer size");
+        return ERANGE;
+    }
+
+    errno = 0;
+    // Returns a pointer to memory that may be statically allocated by glibc.
+    // Data returned by readdir may be overwritten by other readdir calls 
+    // on the same directory stream.
+    dirent* entry = readdir(dir);
+
+    // positive error number returned -> failure
+    if (errno != 0)
```

Comment:
readdir is documented to return non-NULL on success and NULL on either end-of-stream or error.  While unlikely in an implementation, I don't believe it makes any guarantees about errno being 0 in the success case, in which case it's theoretically possible that readdir could return non-NULL but still have a non-0 errno.  As such, I think this and the subsequent condition should be changed to something more like:
```C++
if (entry == nullptr) // error or end-of-stream
{
    *outputEntry = {}; // managed out param must be initialized
    if (errno != 0)
    {
        assert(errno == EBADF); // invalid directory stream description dir
        return errno;
    }
    return -1;
}
```